astra/src/main/java/com/slack/astra/clusterManager/CacheNodeAssignmentService.java

```diff
@@ -0,0 +1,568 @@
+
+    return snapshotIdsToMetadata;
+  }
+
+  /**
+   * Generates a map of replica metadata keyed by snapshot ID. This method takes a list of replica
+   * metadata and maps each snapshot ID to its corresponding replica metadata.
+   *
+   * @param replicaMetadataList the list of replica metadata
+   * @return a map of snapshot IDs to replica metadata
+   */
+  public static Map<String, ReplicaMetadata> replicaMetadataBySnapshotId(
+      List<ReplicaMetadata> replicaMetadataList) {
+    Map<String, ReplicaMetadata> snapshotIdsToMetadata = new HashMap<>();
+    replicaMetadataList.forEach(
+        replicaMetadata ->
+            snapshotIdsToMetadata.putIfAbsent(replicaMetadata.snapshotId, replicaMetadata));
+
+    return snapshotIdsToMetadata;
+  }
+
+  /**
+   * Retrieves a list of snapshot metadata for the given snapshot IDs. This method takes a map of
+   * snapshot metadata and a list of snapshot IDs, and returns a list of snapshot metadata
+   * corresponding to those IDs.
+   *
+   * @param snapshotIdsToMetadata a map of snapshot IDs to snapshot metadata
+   * @param snapshotIds the list of snapshot IDs to retrieve metadata for
+   * @return a list of snapshot metadata corresponding to the given snapshot IDs
+   */
+  private static List<SnapshotMetadata> getSnapshotsFromIds(
+      Map<String, SnapshotMetadata> snapshotIdsToMetadata, List<String> snapshotIds) {
+    List<SnapshotMetadata> snapshots = new ArrayList<>();
+    for (String snapshotId : snapshotIds) {
+      if (snapshotIdsToMetadata.containsKey(snapshotId)) {
+        snapshots.add(snapshotIdsToMetadata.get(snapshotId));
+      }
+    }
+    return snapshots;
+  }
+
+  /**
+   * Calculates the HPA (Horizontal Pod Autoscaler) value based on new assignments. This method
+   * computes the demand factor as the ratio of total bytes requiring assignment to the total
+   * capacity of cache nodes.
+   *
+   * @param newAssignments a map of new assignments keyed by cache node ID
+   * @param totalCacheNodeCapacityBytes the total capacity of all cache nodes in bytes
+   * @return the calculated HPA value
+   */
+  private static double calculateHpaValueFromNewAssignments(
+      Map<String, CacheNodeBin> newAssignments, long totalCacheNodeCapacityBytes) {
+    long totalBytesRequiringAssignment =
+        newAssignments.values().stream()
+            .mapToLong(
+                cacheNodeBin ->
+                    cacheNodeBin.getTotalCapacityBytes() - cacheNodeBin.getRemainingCapacityBytes())
+            .sum();
+
+    return (double) totalBytesRequiringAssignment / totalCacheNodeCapacityBytes;
+  }
+
+  /** Updates or inserts an (ephemeral) HPA metric for the cache nodes. This is NOT threadsafe. */
+  private static void persistCacheConfig(
+      HpaMetricMetadataStore hpaMetricMetadataStore, String replicaSet, Double demandFactor) {
+    String key = String.format(CACHE_HPA_METRIC_NAME, replicaSet);
```

Comment:
Make sure that the other HPA publishing service isn't running at the same time, otherwise they'll end up fighting each other.